src/main/java/com/google/api/gax/bundling/FlowController.java

```diff
@@ -0,0 +1,187 @@
+
+    public ElementCountMaxOutstandingReachedException(int currentMaxElementCount) {
+      this.currentMaxElementCount = currentMaxElementCount;
+    }
+
+    public int getCurrentMaxBundleElementCount() {
+      return currentMaxElementCount;
+    }
+
+    @Override
+    public String toString() {
+      return String.format(
+          "The maximum number of bundle elements: %d have been reached.", currentMaxElementCount);
+    }
+  }
+
+  /**
+   * Returned as a future exception when client-side flow control is enforced based on the maximum
+   * number of unacknowledged in-memory bytes.
+   */
+  public final class RequestByteMaxOutstandingReachedException extends FlowControlException {
+    private final int currentMaxBytes;
+
+    public RequestByteMaxOutstandingReachedException(int currentMaxBytes) {
+      this.currentMaxBytes = currentMaxBytes;
+    }
+
+    public int getCurrentMaxBundleBytes() {
+      return currentMaxBytes;
+    }
+
+    @Override
+    public String toString() {
+      return String.format(
+          "The maximum number of bundle bytes: %d have been reached.", currentMaxBytes);
+    }
+  }
+
+  @AutoValue
+  public abstract static class Settings {
+    public static Settings DEFAULT =
+        newBuilder()
+            .setRequestByteMaxOutstanding(Optional.<Integer>absent())
+            .setElementCountMaxOutstanding(Optional.<Integer>absent())
+            .build();
+
+    /** Maximum number of outstanding elements to keep in memory before enforcing flow control. */
+    public abstract Optional<Integer> getElementCountMaxOutstanding();
+
+    /** Maximum number of outstanding bytes to keep in memory before enforcing flow control. */
+    public abstract Optional<Integer> getRequestByteMaxOutstanding();
+
+    public Builder toBuilder() {
+      return new AutoValue_FlowController_Settings.Builder(this);
+    }
+
+    public static Builder newBuilder() {
+      return new AutoValue_FlowController_Settings.Builder();
+    }
+
+    @AutoValue.Builder
+    public abstract static class Builder {
+      public abstract Builder setElementCountMaxOutstanding(Optional<Integer> value);
+
+      public abstract Builder setRequestByteMaxOutstanding(Optional<Integer> value);
+
+      abstract Settings autoBuild();
+
+      public Settings build() {
+        Settings settings = autoBuild();
+        Preconditions.checkArgument(
+            settings.getElementCountMaxOutstanding().or(1) > 0,
+            "elementCountMaxOutstanding limit is disabled by default, but if set it must be set to a value greater than 0.");
+        Preconditions.checkArgument(
+            settings.getRequestByteMaxOutstanding().or(1) > 0,
+            "requestByteMaxOutstanding limit is disabled by default, but if set it must be set to a value greater than 0.");
+        return settings;
+      }
+    }
+  }
+
+  @Nullable private final Semaphore outstandingElementCount;
+  @Nullable private final Semaphore outstandingByteCount;
```

Comment:
Done.

According to the [Java style guide](https://google.github.io/styleguide/javaguide.html#s4.8.5-annotations), this is OK so the formatter didn't catch it. Annotations on their own lines seems to be more in-line with the rest of the repo though.